### PR TITLE
Add option to train AI using Tasks and their bounding boxes

### PR DESCRIPTION
**Pull request by MichaelBuessemeyer**
This PR changes the AI Training Input under Administration -> AI Models to support task Ids. Additionally, the task bounding boxes of those tasks are now checked before submitting the training job. The [corresponding worker pr](https://github.com/scalableminds/voxelytics/pull/3883) ensures that the worker also uses the task bounding boxes during training. 

### URL of deployed dev instance (used for testing):
- https://___.webknossos.xyz

### Steps to test:
- enable jobs
- optional: create a new task type that allows volume annotation and not only skeleton (as the sample type)
- create two new tasks with a task bounding box each. Ensure that the task bounding box is mag aligned in the mag you want to choose for training. Choosing mag 1 should be easiest. The task bounding boxes should have a size of 64 in each dimension.
- do some volume annotation (withing the task bbox) for these two tasks and then finish each of them.
- create two more annotations. For each create at least one user bounding box of size 64 (mag aligned) and do some volume annotation in these.
- Go to the administration -> ai models view and open the task/annotationID input to start a new ai training.
- Enter the two task ids and the two annotation ids. No error should occur (even though the frontend falsely try whether the annotation ids are actually task ids which results in an error sent by the backend).
- Then proceeded to the modal to configure the training. Choose the volume annotation layers for the annotations (not tasks) and choose the mag you decided for. Then submit.
- Start the local worker.
- The training should successfully start successfully.
- Then try the same again. This time with non-mag aligned bounding boxes. The modal should show a warning that bounding boxes are not mag aligned.

### Issues:
- fixes https://github.com/scalableminds/webknossos/issues/8265
- corresponding worker pr: https://github.com/scalableminds/voxelytics/pull/3883

------
(Please delete unneeded items, merge only when none are left open)
- [x] Updated [changelog](../blob/master/CHANGELOG.unreleased.md#unreleased)

### CODE REVIEW

**Comment by MichaelBuessemeyer** [comment]
@daniel-wer I just noticed that the worker bucket & mag aligns the bounding boxes before using them. This is not done by the frontend checking mechanism. But this aligning correction might lead to different sizes of the bounding boxes and therefore the frontend assertion that bounding boxes have a multiple size of the smallest bounding box might no longer be correct. => Therefore I would also suggest doing the aligning in the frontend and then doing the bbox size assertion checking.

 Moreover, the currently only bboxes with an extent of < 10 in at least one direction are counted as too small. But according to https://scm.slack.com/archives/C03B6160ATY/p1736344267807119 this is too small. Therefore, I would also suggest to increase this limit. But, I am not familiar enough with the training code to tell what would be a better limit. I would need you're help here :pray:

**Comment by MichaelBuessemeyer** [comment]
Moreover, we could also check that enough annotations are provided. When training with multiple annotations, the training code seems to assume that at least 4 annotations are given by the user

**Comment by daniel-wer** [comment]
>  I just noticed that the worker bucket & mag aligns the bounding boxes before using them. This is not done by the frontend checking mechanism. But this aligning correction might lead to different sizes of the bounding boxes and therefore the frontend assertion that bounding boxes have a multiple size of the smallest bounding box might no longer be correct. => Therefore I would also suggest doing the aligning in the frontend and then doing the bbox size assertion checking.

Good catch, please do so if it's not too complicated. Maybe also a warning should be shown if a bounding box is not mag-aligned.  This would make it somewhat easier for users to understand why bounding box sizes might differ in mags > 1.

> Moreover, the currently only bboxes with an extent of < 10 in at least one direction are counted as too small. But according to https://scm.slack.com/archives/C03B6160ATY/p1736344267807119 this is too small. Therefore, I would also suggest to increase this limit. But, I am not familiar enough with the training code to tell what would be a better limit. I would need you're help here 🙏

I think it should be fine to go with a minimum extent of 32 vx in the training mag. Do you agree @valentin-pinkau?

> Moreover, we could also check that enough annotations are provided. When training with multiple annotations, the training code seems to assume that at least 4 annotations are given by the user

Mmh, I'm not sure I get that. I don't know of any minimum number of annotations assumed by the training code. Do you mean bounding boxes? If so, the `test_fraction` config value determines a minimum number of training samples that's needed so that the validation set contains at least one example (e.g. if the test_fraction is 0.2 at least 5 training samples are needed). Or do you mean something else?

**Comment by MichaelBuessemeyer** [comment]
> If so, the test_fraction config value determines a minimum number of training samples that's needed so that the validation set contains at least one example (e.g. if the test_fraction is 0.2 at least 5 training samples are needed). Or do you mean something else?

Yes, I mean exactly that: The training process always died on me before actually starting the training because of a `test_fraction` 0.3. I could only resolve this issue by providing at least 4 annotations and not by adding more bounding boxes to one of the three annotations I previously tried. At least that's what I observed

**Comment by daniel-wer** [comment]
> Yes, I mean exactly that: The training process always died on me before actually starting the training because of a test_fraction 0.3. I could only resolve this issue by providing at least 4 annotations and not by adding more bounding boxes to one of the three annotations I previously tried. At least that's what I observed

I just retested this by adding more and more bounding boxes to a single annotation. Once I reached 4 bounding boxes, the training no longer failed. Alternatively, enlarging a single bounding box also works. Not sure what happened in your case :thinking: 

The test_fraction is computed dynamically, but it is at most 0.3 and only goes down if the bounding boxes contain more than ~16M voxels at which point there should be more training samples than are needed to have at least one example in the validation set. However, what complicates this is that not the number of bounding boxes is critical here, but the number of training samples. Larger bounding boxes can contain more than one training sample. I think it would be very complicated to calculate this in the frontend and would therefore not do anything about it for now. If we see users hitting this issue we can reconsider this decision.

**Comment by MichaelBuessemeyer** [comment]
@daniel-wer Can you do the review of this pr and the corresponding worker pr as you are already familiar with the issue? The worker pr is very small and should be easy to understand.

**Comment by daniel-wer** [comment]
Yes I'll happily do the reviews :)

**Review by daniel-wer — COMMENTED** [review]
Well done :+1: Testing also went well :tada: 

During testing I noticed that the minimum dimension warning does not seem to take the mag into account. At least I had a bounding box with extent 32³, chose mag 2-2-2, and there was no warning that it was too small (although it is only 16³ vx in mag 2-2-2). I think this was wrong before your PR, so you don't need to fix it if it's more involved.

File: frontend/javascripts/oxalis/view/jobs/train_ai_model.tsx [hunk 1]
```diff
@@ -571,14 +622,25 @@ function checkBoundingBoxesForErrorsAndWarnings(
...
   });
 
+  if (notMagAlignedBoundingBoxes.length > 0) {
+    hasBBoxWarnings = true;
+    const notMagAlignedBoundingBoxesStrings = notMagAlignedBoundingBoxes.map(
+      ({ boundingBox, name, annotationId, alignedBoundingBox }) =>
+        `'${name}' of annotation ${annotationId}: ${boundingBox.join(", ")} -> ${alignedBoundingBox.join(", ")}`,
+    );
+    warnings.push(
+      `The following bounding boxes are not aligned with the selected magnification. They will be automatically shrunk to be aligned with the magnification:\n${notMagAlignedBoundingBoxesStrings.join("\n")}`,
```
[the remark this answers is not in the document]
Comment (MichaelBuessemeyer) [reply frontend/javascripts/oxalis/view/jobs/train_ai_model.tsx:632]: Yeah, the problem is that one can select different mags for different annotations. Therefore, I added the mag to the end of each bbox entry in the warning message.
But the problem is, that this makes the text even longer. I don't know if that's a benefit.
Maybe a better tree structure of the error message would help here :thinking:. But this would also complicate the the code a little.

Here is what I currently have:
image

And this is what I mean with an improved hierarchy structure:
```markdown
The following bounding boxes are not aligned with the selected magnification. They will be automatically shrunk to be aligned with the magnification:
- Annotation 677e8b3f5f0100a80b1dc6ae
  - 'Bounding box 1' (3584, 3584, 1024, 64, 64, 64) will be 1792, 1792, 1024, 32, 32, 64 in mag 2, 2, 1
  - 'Bounding box 2' (3648, 3648, 1088, 64, 64, 64) will be 1824, 1824, 1088, 32, 32, 64 in mag 2, 2, 1
  - 'Bounding box 3' (3712, 3712, 1152, 64, 64, 64) will be 1856, 1856, 1152, 32, 32, 64 in mag 2, 2, 1 
```
What do you think about this tree structured text?

[the remark this answers is not in the document]
Comment (daniel-wer) [reply frontend/javascripts/oxalis/view/jobs/train_ai_model.tsx:634]: I like that a lot :+1:

[the remark this answers is not in the document]
Comment (MichaelBuessemeyer) [reply frontend/javascripts/oxalis/view/jobs/train_ai_model.tsx]: Ok, I made it work :tada: 
image

**Review by daniel-wer — APPROVED** [review]
LGTM :+1: 

> During testing I noticed that the minimum dimension warning does not seem to take the mag into account. At least I had a bounding box with extent 32³, chose mag 2-2-2, and there was no warning that it was too small (although it is only 16³ vx in mag 2-2-2). I think this was wrong before your PR, so you don't need to fix it if it's more involved.

Was this fixed by your changes or should it be fixed as a followup?

**Comment by MichaelBuessemeyer** [comment]
> Was this fixed by your changes or should it be fixed as a followup?


That should also be solved here. The screenshot shows that the mag adjusted size is used for minimal bounding box comparison
image

**Comment by MichaelBuessemeyer** [comment]
And this screenshot includes the minimum warning message 
image

**Comment by daniel-wer** [comment]
Awesome, thank you!
